src/decoding/error.rs

```diff
@@ -0,0 +1,119 @@
+use std::{
+    fmt::{self, Display, Formatter},
+    num, str, string,
+};
+
+use failure::Fail;
+
+use crate::state_tracker::StructureError;
+
+#[derive(Debug, Fail)]
+pub struct Error {
+    #[fail(context)]
+    context: Option<String>,
+    #[fail(cause)]
+    error: ErrorKind,
+}
+
+/// An enumeration of potential errors that appear during bencode deserialization.
+#[derive(Debug, Fail)]
+pub enum ErrorKind {
+    /// Error that occurs if the serialized structure contains invalid information.
+    #[fail(display = "malformed content discovered: {}", _0)]
+    MalformedContent(failure::Error),
+    /// Error that occurs if the serialized structure is incomplete.
+    #[fail(display = "missing field: {}", _0)]
+    MissingField(String),
+    /// Error in the bencode structure (e.g. a missing field end separator).
+    #[fail(display = "bencode encoding corrupted")]
+    StructureError(#[fail(cause)] StructureError),
+    /// Error that occurs if the serialized structure contains an unexpected field.
+    #[fail(display = "unexpected field: {}", _0)]
+    UnexpectedField(String),
+    /// Error through an unexpected bencode token during deserialization.
+    #[fail(display = "discovered {} but expected {}", _0, _1)]
+    UnexpectedToken(String, String),
+}
+
+pub trait ResultExt {
+    fn context(self, context: impl std::fmt::Display) -> Self;
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match &self.context {
+            Some(context) => write!(f, "Error: {} in {}", self.error, context),
+            None => write!(f, "Error: {}", self.error),
+        }
+    }
+}
+
+impl From<ErrorKind> for Error {
+    fn from(error: ErrorKind) -> Self {
+        Error {
+            context: None,
+            error,
+        }
+    }
+}
+
+impl Error {
+    pub fn context(mut self, context: impl Display) -> Self {
+        self.context = Some(context.to_string());
+        self
+    }
+
+    /// Raised when there is a general error while deserializing a type.
+    /// The message should not be capitalized and show not end with a period.
```

Comment:
Typo? "show not end" -> "should not end"

Reply:
agreed

Reply:
thx!